test/mpi/util/mtest_thread.c

```diff
@@ -118,9 +119,20 @@ int MTest_thread_barrier(int nt)
     }
     /* Really need a write barrier here */
     *cntP = *cntP - 1;
+    wait = *cntP > 0;
     err = MTest_thread_unlock(&barrierLock);
     CHECK_LOCK(err);
-    while (*cntP > 0);
+
+    /* wait for other threads */
+    while (wait) {
+        err = MTest_thread_lock(&barrierLock);
+        CHECK_LOCK(err);
+
+        wait = *cntP > 0;
+
+        err = MTest_thread_unlock(&barrierLock);
+        CHECK_LOCK(err);
+    }
```

Comment:
What we really needed is an atomic variable. Maybe use `MPL_atomic`?

Reply:
Integrating MPL into the testsuite will take some time, and its not something I wish to backport to 4.0.x just for this fix. If desired we can drop this and only target 4.0.x.

Reply:
I see. Let me ponder it a bit. What was the original issue? Are all thread tests exhibit this instability?

Reply:
Threads read the value outside a critical section while others update it. On a weak memory system is it likely to see corruption.

Reply:
I agree atomics would be an improvement. It's just not trivial given how we do atomics in MPICH.

Reply:
Okay. How about adding a comment explaining "ideally that *cntP should be atomic. Since managing portable atomics is complicated, we use critical sections here instead"?